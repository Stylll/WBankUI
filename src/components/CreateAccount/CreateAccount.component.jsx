import React, { useState, useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as accountActions from '../../redux/actionCreators/accountActions';
import TextInput from '../TextInput/TextInput.component';
import Button from '../Button/Button.component';
import { validateCreateAccount } from '../../helpers/validators/accountValidator';
import './CreateAccount.styles.scss';

const CreateAccount = ({
  createAccount, isLoading: reduxIsLoading, closeModal,
  errorMessage, accountCreateSuccess, dataTestId,
}) => {
  const [values, setValues] = useState({
    name: '',
    openingBalance: '',
  });

  const [formErrors, setFormErrors] = useState({
    name: '',
    openingBalance: '',
  });

  const [serverError, setServerError] = useState('');

  const [isLoading, setIsLoading] = useState(false);

  const formSubmitted = useRef(false);
  const hasServerOperationOccurred = useRef(false);

  useEffect(() => {
    if (!reduxIsLoading && accountCreateSuccess && formSubmitted.current) {
      setIsLoading(false);
      closeModal();
      formSubmitted.current = false;
      return;
    }

    if (!reduxIsLoading && formSubmitted.current) {
      setIsLoading(false);
      formSubmitted.current = false;
    }
  }, [reduxIsLoading, accountCreateSuccess, closeModal]);

  useEffect(() => {
    if (errorMessage && hasServerOperationOccurred.current) {
      setServerError(errorMessage);
    }
  }, [errorMessage]);

  const handleChange = (event) => {
    event.preventDefault();
    setValues({
      ...values,
      [event.target.name]: event.target.value,
    });
  };

  const handleClick = (event) => {
    event.preventDefault();
    const { isValid, errors } = validateCreateAccount(values);
    setFormErrors({
      ...errors,
    });
    if (!isValid) {
      return;
    }

    setIsLoading(true);

    createAccount({
      name: values.name,
      openingBalance: values.openingBalance,
    });

    formSubmitted.current = true;
    hasServerOperationOccurred.current = true;
  };

  const renderCreateForm = () => (
    <div className="create-container">
        <div>
            <p>Create your bank account</p>
            <span className="error-text">{serverError}</span>
            <TextInput
                name="name"
                placeholder="My Savings Account"
                title="Enter your preferred account name"
                type="text"
                value={values.name}
                error={formErrors.name}
                onChange={handleChange}
                dataTestId={`${dataTestId}-name`} />
            <TextInput
                name="openingBalance"
                placeholder="5000"
                title="Enter your Opening balance"
                type="text"
                value={values.openingBalance}
                error={formErrors.openingBalance}
                onChange={handleChange}
                dataTestId={`${dataTestId}-openingBalance`} />
            <Button
                title="Create"
                handleClick={handleClick}
                showLoader={isLoading}
                disabled={isLoading}
                dataTestId={`${dataTestId}-create`}
            />
        </div>
        <br />
        <h4 className="link">
          Please note, opening balance is saved in CAD.
        </h4>
    </div>
  );

  return (
        <div className="create-account-container">
            <div className="form-container">
                    {renderCreateForm()}
                </div>
        </div>
  );
};

CreateAccount.propTypes = {
  createAccount: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  accountCreateSuccess: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
  dataTestId: PropTypes.string,
};

CreateAccount.defaultProps = {
  dataTestId: 'createaccount',
};

const mapStateToProps = ({ account }) => ({
  isLoading: account.isCreateAccountLoading,
  accountCreateSuccess: account.accountCreateSuccess,
  errorMessage: account.errorMessage,
});

const mapDispatchToProps = {
  createAccount: accountActions.createAccount,
};

export {
  CreateAccount,
};

export default connect(mapStateToProps, mapDispatchToProps)(CreateAccount);
