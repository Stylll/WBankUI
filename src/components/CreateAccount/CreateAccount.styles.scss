@import '../../partials/colors';

.create-account-container {
    display: flex;
    justify-content: center;

    .form-container {
        width: 550px;
    }

    .create-container {
        box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
        margin: 40px;
        padding: 40px;
        background-color: $white;

        .link {
            font-weight: 100;
            &:hover {
                text-decoration: underline;
                cursor: pointer;
            }
        }
        .error-text {
            color: $red;
        }
    }
}

/* small to medium media query */
@media only screen and (max-width: 600px) {
    .create-account-container {
        .form-container {
            width: auto;
        }
    }
}
