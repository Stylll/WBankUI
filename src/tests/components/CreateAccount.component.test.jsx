import { render, fireEvent } from '@testing-library/react';
import React from 'react';

import { CreateAccount } from '../../components/CreateAccount/CreateAccount.component';

describe('CreateAccount Test', () => {
  it('should render properly', () => {
    const { baseElement } = render(<CreateAccount
      createAccount={() => {}}
      closeModal={() => {}}
      isLoading={false}
      accountCreateSuccess={false}
      errorMessage=""
    />);
    expect(baseElement).toMatchSnapshot();
  });
  it('should validate the input', () => {
    const { baseElement, getByTestId, getByText } = render(<CreateAccount
      createAccount={() => {}}
      closeModal={() => {}}
      isLoading={false}
      accountCreateSuccess={false}
      errorMessage=""
    />);
    fireEvent.click(getByTestId('createaccount-create'));
    expect(getByText('name cannot be empty')).toBeInTheDocument();
    expect(getByText('opening balance cannot be empty')).toBeInTheDocument();
  });
  it('should submit form and call createAccount and display error', () => {
    const createAccount = jest.fn();
    const {
      baseElement, getByTestId, getByText, rerender,
    } = render(<CreateAccount
      createAccount={createAccount}
      closeModal={() => {}}
      isLoading={false}
      accountCreateSuccess={false}
      errorMessage=""
    />);

    fireEvent.change(getByTestId('createaccount-name'),
      { target: { value: 'My account', name: 'name' } });
    fireEvent.change(getByTestId('createaccount-openingBalance'),
      { target: { value: '245', name: 'openingBalance' } });
    fireEvent.click(getByTestId('createaccount-create'));
    expect(createAccount).toHaveBeenCalled();

    rerender(<CreateAccount
      createAccount={createAccount}
      closeModal={() => {}}
      isLoading={false}
      accountCreateSuccess={false}
      errorMessage="An error occurred"
    />);
    expect(getByText('An error occurred')).toBeInTheDocument();
  });
});
