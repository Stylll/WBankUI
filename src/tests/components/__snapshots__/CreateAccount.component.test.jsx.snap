// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccount Test should render properly 1`] = `
<body>
  <div>
    <div
      class="create-account-container"
    >
      <div
        class="form-container"
      >
        <div
          class="create-container"
        >
          <div>
            <p>
              Create your bank account
            </p>
            <span
              class="error-text"
            />
            <div
              class="form-input"
            >
              <div
                class="title"
              >
                Enter your preferred account name
              </div>
              <input
                class="text-input"
                data-testid="createaccount-name"
                id=""
                name="name"
                placeholder="My Savings Account"
                type="text"
                value=""
              />
              <span
                class="info"
              />
              
            </div>
            <div
              class="form-input"
            >
              <div
                class="title"
              >
                Enter your Opening balance
              </div>
              <input
                class="text-input"
                data-testid="createaccount-openingBalance"
                id=""
                name="openingBalance"
                placeholder="5000"
                type="text"
                value=""
              />
              <span
                class="info"
              />
              
            </div>
            <div
              class="btn-button"
            >
              <button
                class="button "
                data-testid="createaccount-create"
                type="submit"
              >
                Create
              </button>
            </div>
          </div>
          <br />
          <h4
            class="link"
          >
            Please note, opening balance is saved in CAD.
          </h4>
        </div>
      </div>
    </div>
  </div>
</body>
`;
